torch/autograd/functional.py

```diff
@@ -0,0 +1,714 @@
+import torch
+
+# Utility functions
+
+def _as_tuple(inp, arg_name, fn_name):
+    # Ensures that the inputs is a tuple of Tensors
+    # Returns whether or not the original inp was a tuple and the tupled version of the input
+    tuple_inp = True
+    if not isinstance(inp, tuple):
+        inp = (inp,)
+        tuple_inp = False
+
+    for i, el in enumerate(inp):
+        if not torch.is_tensor(el):
+            raise TypeError("The {} given to {} must be either a Tensor or a tuple of Tensors but the"
+                            " value at index {} has type {}.".format(arg_name, fn_name, i, type(el)))
+
+    return tuple_inp, inp
+
+def _tuple_postprocess(inp, tuple_inp):
+    # Unpacks a potentially nested tuple of Tensors
+    # It is used to invert _as_tuple before returning to the user
+    if isinstance(tuple_inp, tuple):
```

Comment:
So `tuple_inp` can be either a boolean or a Pair of boolean? I'm not really sure what it does when it is a tuple.

Reply:
When its a tuple of two values, it applies the logic twice in a nested manner.
Inp is always a tuple of tuple of Tensors in that case, and tuple_inp[0] is used to remove the first level of tuple in inp. And tuple_inp[1] is used to remove the inner tuple of inp.

This is used for Jacobian and hessian.

Reply:
this PR looks pretty strange -- the first return of as_tuple is a boolean called tuple_inp, which is the same as the argument to _tuple_postprocess, but that value is never used as a bool.

Reply:
It is used as a bool for functions that return something that match the shape of a single thing (input or output of the user function).
For results that match the shape of two things (like output + input for jacobian), we could do two calls to `_tuple_postprocess` for each boolean tuple_*, but I though it was clearer to combine them in a tuple and have the function handle the nesting itself.

Is that the answer you're looking for? Or maybe I am not understanding the question?

Reply:
Just pulling on this thread a little: it's confusing that:
- `tuple_inp` is used here and in the previous function (as_tuple), but the tuple_inp here can be different from in as_tuple. The confusing part is that `_tuple_postprocess` is advertised as the inverse of `tuple_inp` (it is not always the inverse, it does something slightly different when `tuple_inp` is a tuple).
- `tuple_inp` has the word "tuple" in it. This makes it hard for the reader to reason about what its type is -- "is it a bool? is it a tuple"? Then the reader realizes that it's either a Tuple[Bool] or a Bool.

It might be clearer to split `_tuple_postprocess` into (1) a function that inverts `_as_tuple` once and (2) some function that undoes the nested _as_tuple, as needed for jacobian and hessian

Reply:
Will the following changes work:
- change `tuple_inp` to `is_inp_tuple`
- advertise `_as_tuple` as ensuring everything is a tuple of Tensor.
- advertise `_tuple_postprocess` as doing unpacking of tuples to match the user's original expectations.